File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/payment/infrastructure/persistence/PaymentMapper.java
package com.example.accountservice.payment.infrastructure.persistence;

import com.example.accountservice.payment.domain.Payment;

/**
 * A class dedicated to converting between Payment (pure domain) and PaymentJpaEntity (JPA mapping).
 * Used only inside PaymentRepositoryImpl — the Domain/Application layers are unaware of this class.
 */
final class PaymentMapper {

    private PaymentMapper() {}

    static Payment toDomain(PaymentJpaEntity entity) {
        return Payment.reconstitute(
                entity.getPaymentId(),
                entity.getCardId(),
                entity.getAccountId(),
                entity.getOwnerId(),
                entity.getAmount(),
                entity.getStatus(),
                entity.getCreatedAt());
    }

    /** Creates a new entity (no PK, an insert target) for a new Payment. */
    static PaymentJpaEntity toNewEntity(Payment payment) {
        return new PaymentJpaEntity(
                null,
                payment.getPaymentId(),
                payment.getCardId(),
                payment.getAccountId(),
                payment.getOwnerId(),
                payment.getAmount(),
                payment.getStatus(),
                payment.getCreatedAt());
    }

    /**
     * Applies the domain Payment's latest state to an existing entity (preserving PK) — an update
     * target.
     */
    static PaymentJpaEntity updateEntity(PaymentJpaEntity entity, Payment payment) {
        entity.applyMutableState(payment.getStatus());
        return entity;
    }
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/payment/infrastructure/persistence/PaymentMapper.kt
package com.example.accountservice.payment.infrastructure.persistence

import com.example.accountservice.payment.domain.Payment

/**
 * The object dedicated to converting between Payment (pure domain) and PaymentJpaEntity (JPA mapping).
 * Used only inside PaymentRepositoryImpl — the Domain/Application layers know nothing about this object.
 */
internal object PaymentMapper {
    fun toDomain(entity: PaymentJpaEntity): Payment =
        Payment.reconstitute(
            paymentId = entity.paymentId,
            cardId = entity.cardId,
            accountId = entity.accountId,
            ownerId = entity.ownerId,
            amount = entity.amount,
            status = entity.status,
            createdAt = entity.createdAt,
        )

    /** Creates a new entity (no PK, an insert target) for a new Payment. */
    fun toNewEntity(payment: Payment): PaymentJpaEntity =
        PaymentJpaEntity(
            id = null,
            paymentId = payment.paymentId,
            cardId = payment.cardId,
            accountId = payment.accountId,
            ownerId = payment.ownerId,
            amount = payment.amount,
            status = payment.status,
            createdAt = payment.createdAt,
        )

    /** Reflects the domain Payment's latest state (status) onto the existing entity (preserving the PK) — an update target. */
    fun updateEntity(
        entity: PaymentJpaEntity,
        payment: Payment,
    ): PaymentJpaEntity {
        entity.status = payment.status
        return entity
    }
}
